subsys/bluetooth/host/att.c

```diff
@@ -303,11 +269,18 @@ static int chan_send(struct bt_att_chan *chan, struct net_buf *buf)
 		/* bt_l2cap_chan_send does actually return the number of bytes
 		 * that could be sent immediately.
 		 */
-		err = att_chan_send_cb(chan, buf, chan_cb(buf), data);
+		bt_att_tx_cb_t cb = chan_cb(buf);
+
+		err = bt_l2cap_chan_send(&chan->chan.chan, buf);
 		if (err < 0) {
 			data->att_chan = prev_chan;
 			atomic_clear_bit(chan->flags, ATT_PENDING_SENT);
 			return err;
+		} else {
+			/* Consider the PDU sent. Notify user. */
+			if (cb) {
+				cb(chan->att->conn, data);
+			}
```

Comment:
This is a change in behavior, right? 
This is calling the callback once it has been enqueued for transmission, rather than when we have received the number of complete packets event for the transmission.

Reply:
Yes. But it should be fine, right?

Reply:
Does the `cb` here trigger a callback in GATT? 

If not, then it is fine. 

My worry is that if the callback triggers a GATT callback, then then that effectively traverse up to the application, and the buffers will not be free'd before the application callbacks are called. 

So if this only affects ATT, then it is fine

Reply:
It will traverse up to the application in the case of response-less ATT commands and ATT notifications. It's true that the buffers likely may not have been freed yet.

But I discount that problem since the previous behavior was malloc-like / unpredictable. Even though the cb came after the net_buf was freed, some other process may have snatched up the buffer in the meantime. And there are almost always other processes. Eg. responding to incoming ATT requests, other L2CAP traffic like SMP, MTU updates, or other ATT client users.

Reply:
> It will traverse up to the application in the case of response-less ATT commands and ATT notifications. It's true that the buffers likely may not have been freed yet.

But wouldn't that also mean that if we do e.g. `bt_gatt_notify_cb`, then the callback will be called immediate, rather than when it has been sent on air? That may affect applications, expecting it to not only come in a different thread, but also with different timing. 

It's not necessarily a problem, but it should probably be mentioned somewhere. 

> Even though the cb came after the net_buf was freed, some other process may have snatched up the buffer in the meantime. 

Wouldn't this also mean that in the case where `CONFIG_BT_ATT_TX_COUNT=1`, then it will always be impossible to do any other GATT operations? For example it's pretty common to do characteristic discovery in the callback for the service discovery callback, but this change seems to (maybe) prevent that.

Reply:
Changed it to call on buffer destroy instead.